src/evaluators/JSXElement.js

```diff
@@ -0,0 +1,370 @@
+    if (value instanceof StringValue) {
+      let text = cleanJSXElementLiteralChild(value.value);
+      if (text === null) {
+        dynamicChildrenLength--;
+        // this is a space full of whitespace, so let's proceed
+        continue;
+      } else {
+        value.value = text;
+      }
+    }
+    lastChildValue = value;
+    CreateDataPropertyOrThrow(realm, array, "" + dynamicIterator, value);
+    dynamicIterator++;
+  }
+  if (dynamicChildrenLength === 1) {
+    return lastChildValue;
+  }
+
+  Set(realm, array, "length", new NumberValue(realm, dynamicChildrenLength), false);
+  return array;
+}
+
+function evaluateJSXAttributes(
+  elementType: BabelNodeJSXIdentifier | BabelNodeJSXMemberExpression,
+  astAttributes: Array<BabelNodeJSXAttribute | BabelNodeJSXSpreadAttribute>,
+  astChildren: Array<BabelNode>,
+  strictCode: boolean,
+  env: LexicalEnvironment,
+  realm: Realm
+) {
+  let attributes = new Map();
+  let children = evaluateJSXChildren(astChildren, strictCode, env, realm);
+  let defaultProps = getDefaultProps(elementType, env, realm);
+
+  // defaultProps are a bit like default function arguments
+  // if an actual value exists, it should overwrite the default value
+  if (defaultProps !== null) {
+    for (let [key, defaultProp] of defaultProps.properties) {
+      if (defaultProp !== undefined && defaultProp.descriptor !== undefined) {
+        let defaultPropValue = defaultProp.descriptor.value;
+
+        if (defaultPropValue instanceof Value) {
+          if (key === "children") {
+            if (children === null) {
+              children = defaultPropValue;
+            }
+          } else {
+            attributes.set(key, defaultPropValue);
+          }
+        }
+      }
+    }
+  }
+  for (let astAttribute of astAttributes) {
+    switch (astAttribute.type) {
+      case "JSXAttribute":
+        let { name, value } = astAttribute;
+
+        invariant(name.type === "JSXIdentifier", `JSX attribute name type not supported: ${astAttribute.type}`);
+        attributes.set(name.name, evaluateJSXValue(((value: any): BabelNodeJSXIdentifier), strictCode, env, realm));
+        break;
+      case "JSXSpreadAttribute":
+        let spreadValue = GetValue(realm, env.evaluate(astAttribute.argument, strictCode));
+
+        if (spreadValue instanceof ObjectValue) {
+          for (let [key, spreadProp] of spreadValue.properties) {
+            if (spreadProp !== undefined && spreadProp.descriptor !== undefined) {
+              let spreadPropValue = spreadProp.descriptor.value;
+
+              if (spreadPropValue instanceof Value) {
+                if (key === "children") {
+                  children = spreadPropValue;
+                } else {
+                  attributes.set(key, spreadPropValue);
+                }
+              }
+            }
+          }
+        } else {
+          invariant(false, "TODO: only ObjectValue JSX Spread Attributes are supported");
```

Comment:
If there's legal code that can trigger this, throw a `FatalError` with a descriptive message instead.